test/integration/tcp_conn_pool_integration_test.cc

```diff
@@ -0,0 +1,187 @@
+#include <list>
+
+#include "envoy/config/bootstrap/v2/bootstrap.pb.h"
+#include "envoy/server/filter_config.h"
+
+#include "test/integration/integration.h"
+#include "test/integration/utility.h"
+#include "test/server/utility.h"
+#include "test/test_common/registry.h"
+#include "test/test_common/utility.h"
+
+namespace Envoy {
+namespace {
+
+std::string tcp_conn_pool_config;
+
+// Trivial Filter that obtains connections from a TCP connection pool each time onData is called
```

Comment:
Optional idea: I'd like to see the basic TCP proxy session use the connection pool even if there's no performance gains yet because connections are never reused.    That way we can eventually add prefetched connections and health checks without a big change-over.  What do you think?

Reply:
Absolutely. I'd like to do that separately, if there's no pressing reason to do it in one go.

Reply:
Cool.  I look forward to the change and I'm all for landing the big chunk of new code separately from a pretty massive functional changes to TCP proxying :-)

Reply:
I just looked through. This all seems reasonable, but my main complete is that it's basically all code almost directly copied from the HTTP connection pool case. How can we share this code in the future? Can we open some follow up issues on things we might like to do here? Some things I can think of:
1) Use this code for tcp_proxy per @alyssawilk (+1 from me)
2) Use this code in redis_proxy (by passing the right LB context I'm guessing this would be very easy)
3) Somehow build the HTTP connection pool on top of this one? (At least the HTTP/1.1 one)?
4) Other ideas?

Reply:
I'm happy to open up tickets to do those and work through them, interleaved with the other stuff I'm working on.

I'm least sure of 2, because I don't really understand the protocol very well, but it doesn't seem like it should be a problem.

I think the trickiest past of an HTTP 1 conn pool based on this TCP conn pool is to make the CodecClient have the same lifetime as the connection.

Reply:
OK. Can we open a tracking ticket for all of this? Also looks like we need a master merge.

Reply:
Sure. I'll make a ticket and take a pass to make sure this is in shape.